atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionManager.java

```diff
@@ -190,20 +192,30 @@ protected boolean shouldStopRetrying(int numTimesFailed) {
     public <T, C extends PreCommitCondition, E extends Exception> T runTaskWithConditionThrowOnConflict(
             C condition, ConditionAwareTransactionTask<T, C, E> task) throws E, TransactionFailedRetriableException {
         checkOpen();
-        OpenTransaction openTransaction;
+        try (OpenTransactionImpl openTransaction =
+                runTimed(() -> Iterables.getOnlyElement(startTransactions(ImmutableList.of(condition))), "setupTask")) {
+            return openTransaction.execute(txn -> task.execute(txn, condition));
+        }
+    }
+
+    @Override
+    public List<OpenTransactionImpl> startTransactions(List<? extends PreCommitCondition> conditions) {
         try {
-            openTransaction = runTimed(
-                    () -> Iterables.getOnlyElement(startTransactions(ImmutableList.of(condition))), "setupTask");
+            return startTransactionsInternal(conditions);
         } catch (Exception e) {
-            condition.cleanup();
+            Closer closer = Closer.create(); // N.B. using closer to run all cleanup tasks even if one cleanup throws
```

Comment:
Yaaas, much better

Reply:
Although, I just started reading startTransactionsInternal, and my immediate thought was "shit we forgot to cleanup conditions when that throws exceptions...

Reply:
What if instead of this split thing, we try something like this:
* StartTransactionsInternal at the beginning creates a ```Closer onExceptionCloser = Closer.create();```

As you acquire resources, you register things with it, so you'd immediately accumulate all of the conditions there.

Then in #startTransactionsImpl, IFF you end up with an exception, you execute the closer.

Reply:
Basically, this is SUCH critical code, that from a reader's perspective, EVERYTIME I see resource acquisition, on the next line I want it being registered with SOMETHING that's responsible for cleaning it up in case of exceptions.